### PR TITLE
SEAB-4269: Configurable SourceFile path restriction

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR implements a compromise solution to SEAB-4269, allowing the webservice to be configured to restrict SourceFile paths,  to/via a regex which is read from the configuration file at webservice launch.  This will allow us to restrict or make emergency adjustments without having to cut a new release.

Two new variables are added to the webservice configuration:
1. `sourceFilePathRegex`: set to the Java regex which a SourceFile path and absolute path must match.  Optional, if not specified, paths are not restricted.
2. `sourceFilePathViolationMessage`: set to the message included with the exception which is thrown if a path does not match the regex.  Optional, if not specified, a generic message is used.

Per consensus, the `sourceFilePathRegex` defaults to the safe regexp (letters, numbers, dot, slash, dash, and underscore), plus space.  The space is a compromise, because recently, a small number of SourceFiles containing a single space, from the the repo `github.com/nf-core/circdna`, were added to prod.

Although, in the absence of other dangerous characters, the virulence of spaces in paths is reduced, there's still room for hackery antics.   For optimal safety, we might consider attempting to convince any purveyors of spaces to use a different character, and to eventually change our policy to _not_ allow spaces.

Path checking can be enabled/disabled via two static methods of `SourceFile`: `restrictPaths` and `unrestrictPaths`.

Paths are checked in `setPath` and `setAbsolutePath`, and if the path does not match the regex, a `CustomWebApplicationException` with code `SC_BAD_REQUEST` is thrown.  The restriction could have been implemented elsewhere in the system, but this particular enforcement location ensures that nothing will slip by (because a call site was overlooked or whatever) and provides feedback on the bad path relatively early in the process.

I did an experiment, and `SourceFiles` that already exist in the database will load even if their paths don't match the regex, and the webservice appears to operate normally when interacting with entries that include them.  In other words, a specified regex will only apply to the creation of **new** `SourceFiles`.  Non-conformant `SourceFiles` may preexist, and it is up to the admin to determine how to manage them.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-4269

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java [hunk 1]
```diff
@@ -148,6 +148,10 @@ public class DockstoreWebserviceConfiguration extends Configuration {
 
     private String checkUrlLambdaUrl;
 
+    private String sourceFilePathRegex;
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java:151]: Is there a reason why these don't have defaults like `dashboard`?
Think the default could just be your original suggestion

**Review by coverbeck — COMMENTED** [review]
Great unit tests!

The threading comment is for an extreme edge case and unlikely to ever happen (in the tiny window where the JVM may not read the latest value, an invalid source path needs to be submitted), but I still lean towards it should be addressed.

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/ConfigurationIT.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        protected void starting(Description description) {
+            System.out.println("Starting test: " + description.getMethodName());
+        }
+    };
+
+    private void before(DropwizardTestSupport<DockstoreWebserviceConfiguration> support) throws Exception {
+        CommonTestUtilities.dropAndRecreateNoTestData(support);
+        support.before();
+        testingPostgres = new TestingPostgres(support);
+        CommonTestUtilities.cleanStatePrivate2(support, false);
```
[the remark this answers is not in the document]
Comment (svonworl) [reply dockstore-integration-testing/src/test/java/io/dockstore/webservice/ConfigurationIT.java]: done

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/ConfigurationIT.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        file.setAbsolutePath("/Dockstore.cwl");
+        SourceFile file2 = new SourceFile();
+        file2.setContent("{}");
+        file2.setType(SourceFile.TypeEnum.CWL_TEST_JSON);
+        file2.setPath(SOURCEFILE_PATH);
+        file2.setAbsolutePath(SOURCEFILE_PATH);
+        api.editHostedWorkflow(hostedWorkflow.getId(), List.of(file, file2));
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (svonworl) [reply dockstore-integration-testing/src/test/java/io/dockstore/webservice/ConfigurationIT.java:108]: Thanks for the tip, Denis!  Since the last of these tests catches the exception, there's enough structural difference that using parameterized tests doesn't result in a big win in this case.   But I'll keep it in mind for future work!

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java [hunk 4]
```diff
@@ -226,6 +234,22 @@ public void setFrozen(boolean frozen) {
...
     }
 
+    private static void checkPath(String path) {
+        if (path != null && pathRegex != null && !pathRegex.matcher(path).matches()) {
+            throw new CustomWebApplicationException(pathViolationMessage, HttpStatus.SC_BAD_REQUEST);
+        }
+    }
+
+    public static void restrictPaths(Pattern newPathRegex, String newPathViolationMessage) {
+        pathRegex = newPathRegex;
```
[the remark this answers is not in the document]
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java:244]: Yes, technically correct, although the chance of a problem is this particular case is essentially zero.
https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4.3
But I added the `volatile`s anyways, why not.
Note: this pattern happens elsewhere in the existing webservice code.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java (hunk 1 above)
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java:151]: I think I'm ok with just using your restrictions and then loosening later if we get the requests. I think you mentioned that we haven't actually run into any problematic paths yet.

I find it's usually easier in these cases to loosen than tighten.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java:244]: Interesting link, I'm not sure though that it offers extra proof as I read it, although it might be too abstract for me to understand correctly:

1. The first part talks about intra-thread, which I think doesn't apply to my example (my example is two threads).
2. The second part talks about sequential consistency, but then goes on to say "If we were to use sequential consistency as our memory model...", which implies they don't use it.

But yes, agreed, the odds are ridiculously slim in the current use case.

**Review by coverbeck — APPROVED** [review]

**Review by denis-yuen — CHANGES_REQUESTED** [review]
re: discussion in https://github.com/dockstore/dockstore/pull/5005#discussion_r918374052

For potential discussion, not having the restriction doesn't seem to match the consensus reached in https://github.com/dockstore/dockstore/pull/4907 

I think the concerns written up in the google doc linked to the SEAB ticket are convincing enough to start with a restrictive approach where we loosen on request (now that it is simple via configuration file) rather than waiting till we have an actual issue. Can discuss more in https://ucsc-cgl.atlassian.net/browse/SEAB-4269

Tagging @david4096 @coverbeck @bvizzier to resolve this discussion since it seems to be dragging on

**Review by david4096 — APPROVED** [review]
A welcome feature, sounds like we may need to allow ` `.

**Review by kathy-t — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java [hunk 5]
```diff
@@ -234,18 +234,18 @@ public void setFrozen(boolean frozen) {
         this.frozen = frozen;
     }
 
-    private static void checkPath(String path) {
+    private static synchronized void checkPath(String path) {
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java:237]: From a performance standpoint, I think I preferred the volatile approach better, even though synchronized is correct. :) But no need to change, and this is correct.

**Review by coverbeck — APPROVED** [review]
